Adding Apptainer/Singularity support

e-kotov opened the pull request:
I have added support for Apptainer (formerly Singularity) containers.
To do this I:
- moved a few helper functions from `dockerize.R` to `installarion.R`, specifically: `.insert_materials_dir()` and `.write_dockerfile()`. Also `.write_dockerfile()` is now more generic `.write_container_file()`, because it serves both `dockerize()` and `apptainerize()`. `.insert_materials_dir()` now also serves both functions and distinguishes between docker and apptainer/singularity containers. It seems safe to do so, all existing tests pass and I also added a separate test for it in `test_apptainerize.R`.
- added `apptainerize()` function (with name variations). I am not sure you want to "pollute" the package with `singularize()` synonyms. Kindly let me know what you think.
- covered `apptainerize()` with formal tests that mirror what you have for `dockerize()' + I did a lot of testing using the actual singularity module on a linux server. All singularity definitions seem to work just fine.
- caching of debian rootfs tar does not work in an easy way for Apptainer/Singularity, as it requires user to have `debootstrap` installed https://apptainer.org/docs/user/latest/appendix.html#id17 therefore, at least for now, I think it is much more user friendly to skip the caching of Debian image and just use the Docker image of debian:eol in the Apptainer/Singularity definition.
- updated README and FAQ vignette
- extensively tested using Linux VPS, all Apptainer/Singularity image definition files that I get out of the new function seem to build without issues.

e-kotov committed:
move and update .insert_materials_dir() in preparation for support for apptainer/singularity

e-kotov committed:
move and rename .write_dockerfile() to .write_container_file() in preparation for Apptainer/Singularity container support

e-kotov committed:
add Apptainer/Singularity container support (no tests yet)

e-kotov committed:
add Apptainer/Singularity helper functions

e-kotov committed:
rename singularity.def to more neutral container.def container definition file name

e-kotov committed:
fixes to 'from scratch' debian. needs to be tested

e-kotov committed:
fix missing Apptainer/Singularity definition keywords/sections

e-kotov committed:
adapt formal tests from dockerize() to apptainerise()

e-kotov committed:
add space before and after Apptainer/Singularity definition sections

e-kotov committed:
remove debian image cache for apptainer, no easy way to do it, better to bootrstrap from docker image

e-kotov committed:
fix tests for apptainer with custom post install

e-kotov committed:
Update README and FAQ vignette with info about Apptainer/Singularity

e-kotov committed:
fix typos in error and warnings messages

chainsawriot commented on R/installation.R:
```diff
@@ -436,3 +436,213 @@ dockerise <- function(...) {
+#' @param output_dir character, where to put the Apptainer/Singularity definition file and associated content
+#' @param materials_dir character, path to the directory containing additional resources (e.g. analysis scripts) to be copied into `output_dir` and in turn into the Apptainer/Singularity container
+#' @param post_installation_steps character, additional steps to be added before the in the end of `%post` section the Apptainer/Singularity definition file, see an example below
+#' @param image character, which versioned Rocker Docker image to use. Can only be "r-ver", "rstudio", "tidyverse", "verse", "geospatial"
```
L465 to L472 are the same as `dockerize`; it is better to delete them and add

```r
@inheritParams dockerize
```

chainsawriot commented on R/installation.R:
```diff
@@ -436,3 +436,213 @@ dockerise <- function(...) {
+  }
+  ## This should be written in the root level, not base_dir
+  .write_container_file(apptainer_content, file.path(output_dir, "container.def"))
+  if (isTRUE(insert_readme)) {
```
It would be pretty odd to insert a docker readme to the apptainer. Without preparing a new readme for now, maybe it is a better idea to remove this (as well as the `insert_readme` argument)?

e-kotov replied:
I will add a readme for Apptainer/Singularity, forgot about it.

chainsawriot commented on README.md:
```diff
@@ -194,6 +191,80 @@ docker run -p 8787:8787 -e PASSWORD=abc123 --rm --name "rangtest" -ti rang
 With any browser, go to: `local:8787`. The default username is
 `rstudio`, password is as specified.
 
+### Using Apptainer/Singularity containers for HPC
```
To me, the "for HPC" part is not needed.

e-kotov committed:
add tests for new version of .insert_materials_dir

chainsawriot commented on vignettes/faq.Rmd:
```diff
@@ -206,3 +206,45 @@ And launch a container the same way
+
+# Apptainer/Singularity questions
+
+**AQ1: I am on HPC and I don't have Docker there. Can I use Apptainer/Singularity instead of Docker?**
```
Ditto my another comment on HPC.

e-kotov committed:
add missing newline at end of file

chainsawriot reviewed (commented):
Thanks a lot again! These comments are some suggestions, mostly stylistic. I still need to conduct some smoke tests; and will report to you again.

(If you can, please also set the indentation to 4 spaces. Unfortunately, RStudio doesn't support the `.editorconfig` file provided. If you are using RStudio, you might add your `.Rproj`.)

e-kotov committed:
add reference to apptainerize() in export_rang() docs

e-kotov committed:
update readme, remove HPC references

e-kotov committed:
(1) make 4 spaces indent; (2) rewrite .generate_docker_readme() to generate_container_readme() to work with both types of containers; (3) apptainerize() inherits duplicate params from dockerize()

e-kotov committed:
update docs

e-kotov committed:
rename docker readme template, apptainer readme will be added as a separate file

e-kotov committed:
create Rproj file with deafult for 4 spaces for tab

chainsawriot commented on README.md:
````diff
@@ -194,6 +191,80 @@ docker run -p 8787:8787 -e PASSWORD=abc123 --rm --name "rangtest" -ti rang
+``` bash
+cd ~/rocker_test
+sudo apptainer build container.sif container.def
+apptainer instance start container.sif rangtest
````
I suggest

```bash
apptainer run container.sif R
```

chainsawriot commented on R/apptainer.R:
```diff
@@ -0,0 +1,78 @@
+    apptainer_content$FILES <- append(apptainer_content$FILES,
+                                        c(paste0("cache/rpkgs ", file.path(cache_path, "rpkgs")),
+                                          paste0("cache/rsrc ", file.path(cache_path, "rsrc"))))
+    apptainer_content$POST <- prepend(apptainer_content$POST, paste0("export CACHE_PATH=", cache_path))
```
This line gives me an error. Do you really want to use `prepend` (?`rlang::prepend`)?

chainsawriot replied:
Suggest:

```r
apptainer_content$POST <- append(paste0("export CACHE_PATH=", cache_path), apptainer_content$POST)

```

e-kotov replied:
I accept this suggestion. This was `purrr::prepend()` which is (a) deprecated and (b) potentially adds unnecessary dependencies. Worked for me, as I had accidentally loaded `tidyverse` when testing.

chainsawriot commented:
Smoke

- [x] `resolve("github::cran/psych", snapshot_date = "2013-01-01")` (if `prepend` fixed)
- [x] `resolve("chimeraviz", snapshot_date = "2023-01-01")`
- [x] `resolve("chimeraviz", snapshot_date = "2023-01-01")` + quarto
- [x] `resolve("chimeraviz", snapshot_date = "2023-01-01")` + material (although the `materials_dir` seems to work differently to docker)
- [x] `resolve(pkgs = "quanteda", snapshot_date = "2018-10-06")` + rstudio

chainsawriot commented:
@e-kotov Did you try to apptainerize with `image = "rstudio"`? I can't make it work.

```r
somewhere <- ""
x <- resolve(pkgs = "quanteda", snapshot_date = "2018-10-06")
apptainerize(x, somewhere, image = "rstudio")
```

It can be built fine.

```sh
cd somewhere
apptainer build container.sif container.def
```

But can't be launched correctly. Specifically `/init` won't run.

```sh
 ## wouldn't run, with error
apptainer instance start container.sif
## can access via 0.0.0.0:8787, but error
apptainer exec container.sif /usr/lib/rstudio-server/bin/rserver --auth-none=1 --auth-pam-helper-path=pam-helper 
## password won't pass
PASSWORD='abc123' singularity exec container.sif /usr/lib/rstudio-server/bin/rserver --auth-none=0 --auth-pam-helper-path=pam-helper 
```

But I must admit I am a pretty lousy apptainer user (just learned to use it 72 hours ago). Maybe you would have a better idea how it works.

e-kotov commented:
@chainsawriot I have tested, but in a slightly different environment. I am woking on (a) fixes and (b) readme for apptainer. Some more testing needs to be done, as it does not work in every environment reliably yet. I have addressed most of your comments already, but I have not pushed my commits to GitHub yet. I will let you know once I re-check everything so that you don't have to spend your time troubleshooting.

Thanks for your feedback and patience.

e-kotov committed:
Add Apptainer/Singularity README template

e-kotov committed:
fix Apptainer/Singularity startscript, as /init uses s6 overlay seemingly incompatible with Apptainer/Singularity

e-kotov committed:
Update README.md/Rmd and faq.Rmd with correct instructions to start RStudio IDE in Apptainer/Singularity

e-kotov committed:
remove deprecated purrr::prepend()

e-kotov committed:
add info that Apptainer/Singularity containers with RStudio IDE cannot be started using `root` user

e-kotov committed:
update README: (1) compatibility with old RStudio IDE (<4.0) that requires binding .rstudio to the container (2) add default port parameter so that it is obvious how to change it

e-kotov committed:
Fix RStudio Apptainer/Singularity instance start. Add ENV variables that are needed for start of the instance, but not the install process

e-kotov committed:
Update README, readme template for Apptainer/Singularity and FAQ with instructions in running containers

e-kotov committed:
Update test for correct Apptainer/Singularity container for RStudio image

e-kotov commented:
> @e-kotov Did you try to apptainerize with `image = "rstudio"`? I can't make it work.

I did some more testing and here are the details on how to run those. I am also adding the info below to package readme, Apptainer readme template and FAQ vignette.

> To run RStudio IDE in Apptainer/Singularity container, some writeable folders and a config file have to be created locally:
> 
> ```bash
> mkdir -p run var-lib-rstudio-server .rstudio
> printf 'provider=sqlite\ndirectory=/var/lib/rstudio-server\n' > database.conf
> ```
> 
> After that, you can run the container (do not run as `root` user, otherwise you will not be able to login to RStudio IDE).
> 
> Start instance (on default RSTUDIO port 8787):
> 
> ```bash
> apptainer instance start \
>     --bind run:/run,var-lib-rstudio-server:/var/lib/rstudio-server,database.conf:/etc/rstudio/database.conf,.rstudio:/home/rstudio/.rstudio/ \
>     container.sif \
>     rangtest
> ```
> 
> Now open a browser and go to localhost:8787.
> The default username is your local username, default password is 'set_your_password' (if you are using container generated by rang).
> 
> 
> List running instances:
> 
> ```bash
> apptainer instance list
> ```
> 
> Stop instance:
> 
> ```bash
> apptainer instance stop rangtest
> ```
> 
> Start instance with custom port (e.g. 8080) and password:
> 
> ```bash
> apptainer instance start \
>     --enc RPORT=8080
>     --env PASSWORD='set_your_password' \
>     --bind run:/run,var-lib-rstudio-server:/var/lib/rstudio-server,database.conf:/etc/rstudio/database.conf,.rstudio:/home/rstudio/.rstudio/ \
>     container.sif \
>     rangtest
> ```
> 
> Run container with custom `rserver` command line:
> 
> ```bash
> apptainer exec \
>     --env PASSWORD='set_your_password' \
>     --bind run:/run,var-lib-rstudio-server:/var/lib/rstudio-server,database.conf:/etc/rstudio/database.conf,.rstudio:/home/rstudio/.rstudio/ \
>     container.sif \
>     /usr/lib/rstudio-server/bin/rserver \
>     --auth-none=0 --auth-pam-helper-path=pam-helper \
>     --server-user=$(whoami) --www-port=8787
> ```
> 
> If you run the container using `apptainer exec` command, you will have to kill the `rserver` process manually or Cmd/Ctrl+C from the running container to stop the server.
> 
>

e-kotov committed:
CACHE_PATH should be available during container build, not in the enviroment of a running container

e-kotov commented:
@chainsawriot I have addressed all of your review suggestions and questions and have done some more testing on a VPS with both admin and user privileges. I hope my README additions will easily allow you to check the updates and successfully run an Apptainer container.

e-kotov commented:
For quicker tests, I use:
```
somewhere <- ""
x <- resolve(pkgs = "data.table", snapshot_date = "2019-10-06")
apptainerize(x, somewhere, image = "rstudio")
```
It is not 'foolproof', but for quick iteration it is ideal, as data.table has no dependencies and builds very quickly during container builds.

e-kotov committed:
fix typos

e-kotov committed:
remove tests from test_appteinerize.R that have nothing to do with appteinerize() and are carbon copies of tests in tests_dockerize.R

chainsawriot commented:
@e-kotov Thanks a lot! I can get all smoke tests running. I will merge it now. I think there are still some minor issues related to the documentation. But those can be fixed later.

Again, thanks a lot for your tremendous contribution to this package!

e-kotov commented:
@chainsawriot Thank you too! I will look out for issues on your repo related to Apptainer/Singularity.
